ui/components/ui/box/box.js

```diff
@@ -104,54 +88,112 @@ export const ValidBorderColors = [
+  return (
+    typeof value === 'number' ||
+    ((type === 'margin' ||
+      type === 'margin-top' ||
+      type === 'margin-right' ||
+      type === 'margin-bottom' ||
+      type === 'margin-left') &&
+      value === 'auto')
+  );
 }
 
-function generateSizeClasses(baseClass, type, main, top, right, bottom, left) {
-  const arr = Array.isArray(main) ? main : [];
-  const singleDigit = Array.isArray(main) ? undefined : main;
-  if (Array.isArray(main) && ![2, 3, 4].includes(main.length)) {
-    throw new Error(
-      `Expected prop ${type} to have length between 2 and 4, received ${main.length}`,
-    );
-  }
+function isValidString(type, value) {
+  return typeof type === 'string' && typeof value === 'string';
+}
 
-  const isHorizontalAndVertical = arr.length === 2;
-  const isTopHorizontalAndBottom = arr.length === 3;
-  const isAllFour = arr.length === 4;
-  const hasAtLeastTwo = arr.length >= 2;
-  const hasAtLeastThree = arr.length >= 3;
+/**
+ * Generate classnames
+ * Generates classnames for different utility styles
+ * Also accepts responsive props in the form of an array
+ * Maps responsive props to mobile first breakpoints
+ *
+ * @param {string} baseClass - The root or base class name
+ * @param {string} type - The style declaration type "margin", "margin-top", "padding", "display" etc
+ * @param {array || number || string} value - prop value being passed in array props are responsive props
+ * @param {*} validatorFn - The validation function for each type of value
+ * @returns
+ */
 
-  return {
-    [`${baseClass}--${type}-${singleDigit}`]: isValidValue(type, singleDigit),
-    [`${baseClass}--${type}-top-${top}`]: isValidValue(type, top),
-    [`${baseClass}--${type}-right-${right}`]: isValidValue(type, right),
-    [`${baseClass}--${type}-bottom-${bottom}`]: isValidValue(type, bottom),
-    [`${baseClass}--${type}-left-${left}`]: isValidValue(type, left),
-    // As long as an array of length >= 2 has been provided, the first number
-    // will always be for the top value.
-    [`${baseClass}--${type}-top-${arr?.[0]}`]: hasAtLeastTwo,
-    // As long as an array of length >= 2 has been provided, the second number
-    // will always be for the right value.
-    [`${baseClass}--${type}-right-${arr?.[1]}`]: hasAtLeastTwo,
-    // If an array has 2 values, the first number is the bottom value. If
-    // instead if has 3 or more values, the third number will be the bottom.
-    [`${baseClass}--${type}-bottom-${arr?.[2]}`]: hasAtLeastThree,
-    [`${baseClass}--${type}-bottom-${arr?.[0]}`]: isHorizontalAndVertical,
-    // If an array has 2 or 3 values, the second number will be the left value
-    [`${baseClass}--${type}-left-${arr?.[1]}`]:
-      isHorizontalAndVertical || isTopHorizontalAndBottom,
-    // If an array has 4 values, the fourth number is the left value
-    [`${baseClass}--${type}-left-${arr?.[3]}`]: isAllFour,
-  };
+function generateClassNames(baseClass, type, value, validatorFn) {
```

Comment:
This function probably needs the most scrutinizing. It is what creates the classNames from both the conventional singular props and responsive array props.

Reply:
cc @brad-decker @adonesky1

Reply:
Lets add some memoization to this method using lodash, so that if  called repetitively with the same input it doesn't result in recomputation.

Reply:
Updated: https://github.com/MetaMask/metamask-extension/pull/15106/commits/28ee7869a95fddfb10c8597c7805d9cde3782e06

I'm not exactly sure what the resolver function should be returning? In the example I followed they used JSON.stringify to return a string? Still seems to work if I return an array?